tooling/migrate-tags/utils.ts

```diff
@@ -0,0 +1,194 @@
+import _ from "lodash"
+
+// NOTE: old tag is `category: string + selected: string[]`
+export interface LegacyTag {
+  category: string
+  selected: string[]
+}
+
+type TagCategories = NonNullable<CollectionPagePageProps["tagCategories"]>
+
+interface ChildrenWithTags {
+  draftBlobContent: PrismaJson.BlobJsonContent | null
+  publishedBlobContent: PrismaJson.BlobJsonContent | null
+  id: string
+  requiresPublish: boolean
+  content: PrismaJson.BlobJsonContent
+  blobId: string | null
+  draftBlobId: string | null
+}
+export const getChildItemsWithTags = async (
+  parentId: string,
+): Promise<ChildrenWithTags[]> => {
+  return db
+    .selectFrom("Resource")
+    .leftJoin("Version", "Version.id", "Resource.publishedVersionId")
+    .leftJoin("Blob as draftBlob", "draftBlob.id", "Resource.draftBlobId")
+    .leftJoin("Blob as publishedBlob", "publishedBlob.id", "Version.blobId")
+    .select((eb) => {
+      return [
+        "draftBlob.content as draftBlobContent",
+        "publishedBlob.content as publishedBlobContent",
+        // NOTE: If the published version has tags, we need to also
+        // update it in place so that we don't have a long tail
+        // and we can remove the outdated code
+        sql<boolean>`jsonb_array_length("publishedBlob"."content" -> 'page' -> 'tags') > 0`.as(
+          "requiresPublish",
+        ),
+        "Resource.id",
+        eb.fn
+          // NOTE: select draft blob preferentially
+          .coalesce(
+            sql<PrismaJson.BlobJsonContent>`"draftBlob"."content"`,
+            sql<PrismaJson.BlobJsonContent>`"publishedBlob"."content"`,
+          )
+          .as("content"),
+        // NOTE: can update `draftBlobId` (if it exists) directly
+        // but we need to create a new `draftBlob` if `draftBlobId`
+        // doesn't exist
+        "blobId",
+        "draftBlobId",
+      ]
+    })
+    .where((eb) =>
+      eb.or([
+        sql<boolean>`jsonb_array_length("draftBlob"."content" -> 'page' -> 'tags') > 0`,
+        sql<boolean>`jsonb_array_length("publishedBlob"."content" -> 'page' -> 'tags') > 0`,
+      ]),
+    )
+    .where("parentId", "=", parentId)
+    .execute()
+}
+
+export const migrateTags = (collatedTags: {
+  categories: Set<string>
+  mappings: Record<string, Set<string>>
+}) => {
+  const labelToId: Record<string, UUID> = {}
+
+  const tagCategories: TagCategories = _.entries(collatedTags.mappings).map(
+    ([categoryLabel, categoryOptions]) => {
+      const categoryId = randomUUID()
+
+      const options = Array.from(categoryOptions.values()).map((option) => {
+        const id = randomUUID()
+        const label = option
+        labelToId[label] = id
+
+        return {
+          id,
+          label,
+        }
+      })
+
+      return { id: categoryId, options, label: categoryLabel }
+    },
+  )
+
+  return {
+    tagCategories,
+    labelToId,
+  }
+}
+
+export async function getCollectionsOfSiteWithTags(siteId: number) {
```

Comment:
comment: technically this isn't quite `collections` but more any `Resource` that is referenced via `parentId` from a `Resource` with `tags` -> this assumption isn't captured anywhere, so if we have a folder + pages with tags (entirely possible, migrators missed out adding index page), what will happen is that the folder's index pages and child pages will be updated to contain `tagCategories`/`tagged` in a n**on-destructive fashion**